### PR TITLE
libsupport Cache provides an LRU cache 

### PR DESCRIPTION
**Pull request by witchel**
Support for caching properties is split into a different PR because it is messy.

### CODE REVIEW

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [inline libtsuba/include/tsuba/Cache.h]: It was a bummer to me that I had to use the `namespace std` trick below.

**Comment by witchel** [comment]
I'm sorry, I don't like what I've written.  I used a parallel hash map even though the number of entries and their size is small, and there is not multi-threaded access.  And my lock ordering is inconsistent.  So please hold off with the reviews!

**Comment by witchel** [comment]
> I'm sorry, I don't like what I've written. I used a parallel hash map even though the number of entries and their size is small, and there is not multi-threaded access. And my lock ordering is inconsistent. So please hold off with the reviews!

Ok, back in business.  Please review my single-threaded cache.

File: libsupport/CMakeLists.txt
Comment (ddn0) [inline libsupport/CMakeLists.txt]: Is this comment actually true?

The documentation of

https://cmake.org/cmake/help/latest/module/CheckIncludeFileCXX.html

says that the path will be stored in `phmap_FOUND` so I'd assume we'd use this path when setting up our interface library:

https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries

E.g.,

```
add_library(phmap::phmap INTERFACE IMPORTED)
target_include_directories(phmap::phmap INTERFACE ... dir of ${phmap_FOUND} ...)
```

File: libsupport/include/katana/CompilerSpecific.h
Comment (ddn0) [inline libsupport/include/katana/CompilerSpecific.h]: can we delete the copy of this file in `libgalois` then?

File: libsupport/include/katana/ConcurrentMap.h
Comment (ddn0) [inline libsupport/include/katana/ConcurrentMap.h]: nit style: "ConcurrentMap is a concurrent flat hashmap with spin locks.\n\nThe implementation will...."

- Also there is no galois's SimpleLock as it has been moved to support already

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: nit: `lru_iterator` or just `list_it` or `ptr`?

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: PascalCase

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: `const`?

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: nit/hint: the default value of `std::optional<T>` is always `std::nullopt`

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: What is this used for?

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: FWIW, I'd call this `LRUCache` as it doesn't really support general caching policies.

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: Consider some moves here to avoid copying. Though overall, the cache interface seems to assume copying is cheap. Is this intentional?

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: It doesn't look like `MapValue` or `MapType` need to be public. Can we make them private instead?

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: Why not just template on the key type as well? and thus support any hashable thing?

**Review by ddn0 — COMMENTED** [review]
why not move the test of the cache to this repo as well?

File: libsupport/include/katana/CompilerSpecific.h
Comment (witchel) [reply libsupport/include/katana/CompilerSpecific.h]: Whoops, that was my intent.

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: Donaldwhydoyoumakesuchrequests?

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: I literally stared at `MapType `for 5 minutes thinking, "type names are supposed to be PascalCase."  It is weird how resistant my neurons are at adapting to our naming conventions.

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: I thought about doing that, but then thought why not leave it generic.  But sure, I can make it `LRUCache`.

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: I added some std::move's because I'm bad at knowing when those will be important.

I have this comment now at the top of the file.

```
// This is single threaded only, it is not intended to store large objects,
// but rather metadata (e.g., a shared_ptr to a property column).
```

**Comment by witchel** [comment]
> why not move the test of the cache to this repo as well?

I just saw that there was no test directory in the open repo, only in enterprise.  I thought that meant we consider tests enterprise-y.  I think all of the `tsuba_` binaries make more sense in open as they use the public interfaces.  Should I at least relocate my cache test to open?

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: What I wanted to say is, "your key needs to be able to hash itself."  I thought a pure template would not do that, so I provided a base class for any key class.  It seemed kind of like a C++-20 concept, but I'm not sure of that either.

What is the best way to say that I need to know if your key is for node or edge, it should have a name (that will be unique for nodes and unique for edges) and I need to know how to hash it.

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: Hmm, I started doing this and I don't like it.  The replacement policy feels like an implementation detail that should stay out of the name.

File: libtsuba/include/tsuba/Cache.h
Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: I think this issue is covered more completely below, but I need three things from the key, one of which is whether it is a node property or an edge property.

File: libsupport/CMakeLists.txt
Comment (witchel) [reply libsupport/CMakeLists.txt]: Ok.  I just copied that code from the last `CMakeLists.txt`.  But that is no excuse.

@tylershunt might have some insight here, but if I engage my pattern matching "skills" can I simplify this whole rigamarole to this?
```
find_package(phmap REQUIRED)
target_link_libraries(katana_support PUBLIC phmap::phmap)
```

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [inline libtsuba/include/tsuba/Cache.h]: This comment seems somewhat misplaced as the current implementation is not multithreaded, so why even mention that you were unable to make a multithreaded implementation?

Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: It is a "notes for future me/maintainer" sort of thing.  I know in 6 months I'm going to wonder why I didn't just do this as a multi-threaded cache.

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [reply libtsuba/include/tsuba/Cache.h]: That may be true, but the implementation of different policies will lead to varying fields in the implementation class and the current map + linked list implementation may not be suitable for let's say prioritized replacement.

If you want a general interface you try a cache base class, but as you note, the current interface is only appropriate for things that are conveniently copyable. And with a base class, you'd have a hard time expressing the template Value and possible Key parameter.

With templates, you can only achieve textual/static polymorphism (i.e., if it happens to have the same name functions and the return values are interchangable, you can replace one cache class for another) versus the dynamic polymorphism of virtual. The way you indicate substitutability with templates (c.f. virtual base class) is simply to have multiple classes with the same function name and signature (C++ concepts just formalizes this). 

So, with templates the way you indicate there are bunch of interchangable cache classes is simply by convention.

```cpp

template<typename T>
class LRUCache {
public:
  T Get();
};

template<typename T>
class FancyCache {
public:
  T Get();
};

template <typename T>
using MyCache = FancyCache;

MyCache cache;
```

We can swap out cache implementations because they all have the same function names.

vs

```cpp
class Cache {
public:
  virtual Object Get() = 0;
};

class LRUCache: public Cache {
public:
  Object Get() override {};
};

Cache cache = new LRUCache();
```

it is not advisable to mix these two paradigms.

Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: Interesting and appreciated.   I have tried to implement the template form of polymorphism for the Cache Key.

Some replacement policies might not work with the current structure, but some, like MRU or segmented LRU will.

File: libtsuba/include/tsuba/Cache.h
Comment (ddn0) [reply libtsuba/include/tsuba/Cache.h]: I'm still not sure why this is here as the Cache class doesn't really care what the key is as long as it is hashable. `KeyBase` sounds like something that goes with the code making the Cache not here where we define the Cache class for everyone.

All your Key type needs to do is implement the Hash protocol: https://en.cppreference.com/w/cpp/utility/hash

Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: It was still there because I thought it needed to be there, but you are right it doesn't.

I have removed it, but a consequence is that I have to "inject" its hash function into namespace `std`.  That has two disadvantages to my eyes: (1) the fields of the key have to be public so the hash function can read them and (2) it feels unencapsulated to me.  I wish the hash function could be defined as part of the object itself.

Neither of these issues is large, but I would be interested to hear your perspective.

Comment (ddn0) [reply libtsuba/include/tsuba/Cache.h]: > I have removed it, but a consequence is that I have to "inject" its hash function into namespace std. That has two disadvantages to my eyes: (1) the fields of the key have to be public so the hash function can read them

You can either make your std::hash specialization a friend of your key class or just have it call a function you want on the key class itself that then returns the hash.

> and (2) it feels unencapsulated to me. I wish the hash function could be defined as part of the object itself.

You can make the hash function part of the object itself. You just need to tell the standard library what that function is called and you do that by extending `std::hash` as described.

Comment (witchel) [reply libtsuba/include/tsuba/Cache.h]: Ah hah. I like the idea of making it a function called by `std::hash`.  That wasn't the pattern used on the cppreference website, though I should have thought of it.

**Comment by ddn0** [comment]
> I just saw that there was no test directory in the open repo, only in enterprise. I thought that meant we consider tests enterprise-y. I think all of the tsuba_ binaries make more sense in open as they use the public interfaces. Should I at least relocate my cache test to open?

I don't understand. There a several tests in this repo and you even wrote one of them: https://github.com/KatanaGraph/katana/blame/master/libsupport/test/uri.cpp

**Comment by witchel** [comment]
> > I just saw that there was no test directory in the open repo, only in enterprise. I thought that meant we consider tests enterprise-y. I think all of the tsuba_ binaries make more sense in open as they use the public interfaces. Should I at least relocate my cache test to open?
> 
> I don't understand. There a several tests in this repo and you even wrote one of them: https://github.com/KatanaGraph/katana/blame/master/libsupport/test/uri.cpp

Oh, I am using this cache in `libtsuba`, so I was thinking I should test it there (the test is currently in `libtsuba-fs` in enterprise and there is no test directory for `libtsuba `in open).  But my bad, I'll move it to `libsupport`.

**Comment by witchel** [comment]
Can I get a Hallelujah?

**Comment by witchel** [comment]
> Can I get a Hallelujah?

In this case, that means a reviewer approval.  I left the parallel hashmap in enterprise for now.

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h]: why is this included here?

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h:38]: It is somewhat awkward to support all arguments when they are only valid for some cases.

Either:
- Create separate constructors for the Vanilla LRU and the Bytes LRU, e.g., `Cache(capacity, evict_cb)` and `Cache(capacity, evict_cb, value_to_bytes)` as the bytes policy requires `value_to_bytes`
- Create separate factory methods `template<typename Key, typename Value> static Cache<Key,Value> MakeFoo(...)`

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h]: I don't know how I feel about having separate `size` and `bytes` and "capacity in size" and "capacity in bytes" methods. If  the Cache policy is an actual policy, the interface of Cache objects should not change when the policy changes. If there is reason for clients to care that they have a bytes-based Cache versus a item-based Cache, that indicates that the interface itself should be different (i.e., different classes) even if the implementation is largely similar.

Either:
- Make size and capacity report values in units appropriate for the cache policy (i.e., items for `kLRUSize` or bytes in `kLRUBytes`)
- Make two separate classes `LRUCache` and `LRUBytesCache` and optionally use private inheritance or composition to pull in the common implementation

**Review by arthurp — COMMENTED** [review]
I have nothing to add. @ddn0 hit everything I would say.

**Comment by witchel** [comment]
> I have nothing to add. @ddn0 hit everything I would say.

Whew!

File: libsupport/include/katana/Cache.h
Comment (witchel) [reply libsupport/include/katana/Cache.h]: Mistake. It is removed.

File: libsupport/include/katana/Cache.h
Comment (witchel) [reply libsupport/include/katana/Cache.h:38]: I have done option 1.

Comment (witchel) [reply libsupport/include/katana/Cache.h:38]: Oh, I have not.  Hold on!

Comment (witchel) [reply libsupport/include/katana/Cache.h:38]: Ok, sorry, good to go.

File: libsupport/include/katana/Cache.h
Comment (witchel) [reply libsupport/include/katana/Cache.h]: Gone with option 1.

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h]: - Perhaps swap the order of `evict_cb` and `value_to_bytes` to allow either policy/constructor to leave `evict_cb` optional.
- Add Doxygen comments on constructors indicating the reason for two different constructors.

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h:57]: This assert seems partially superfluous with the new constructor

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h:62]: `const`?

File: libsupport/include/katana/Cache.h
Comment (ddn0) [inline libsupport/include/katana/Cache.h:70]: `const`?

**Review by ddn0 — APPROVED** [review]

File: libsupport/include/katana/Cache.h
Comment (witchel) [reply libsupport/include/katana/Cache.h]: Done.

File: libsupport/include/katana/Cache.h
Comment (witchel) [reply libsupport/include/katana/Cache.h:57]: Sure is, good catch.

File: libsupport/include/katana/Cache.h
Comment (witchel) [reply libsupport/include/katana/Cache.h:62]: Sorry, I don't get it.  It is a `const `method.

File: cmake/Modules/KatanaBuildCommon.cmake
Comment (tylershunt) [inline cmake/Modules/KatanaBuildCommon.cmake]: I don't think this is used anywhere in this PR (anymore).

**Review by tylershunt — APPROVED** [review]
LGTM

File: cmake/Modules/KatanaBuildCommon.cmake
Comment (witchel) [reply cmake/Modules/KatanaBuildCommon.cmake]: It isn't, but I thought it was useful.  I'll take it out.

**Pull request merged**
